[EXP][Command-Buffer] Support for using copy queue

mfrancepillois opened the pull request:
- Add another command-lists for copy operations.
- If a main copy engine is available on the device, memory operations are enqueued in the *copy* command-list while 
the other commands are enqueued in the general command-list.
- On submission, if not empty, the *copy* command-list is sent to the main copy command-queue while the general command-list is sent to the compute command-queue.
- Both are executed concurrently. Synchronization between the command-lists is handled by Level-Zero events.
- Because of this the optimization is not currently available with in-order command-lists.
- This functionality respects the following existing L0 environment variables:
  - `UR_L0_USE_COPY_ENGINE_FOR_FILL`
  - `UR_L0_USE_COPY_ENGINE_FOR_D2D_COPY`

mfrancepillois commented:
DPC++ linked PR: https://github.com/reble/llvm/pull/362

Commit:
[CTS] Add CTS tests for urMemImageCreate entry-point

* Adds CTS tests for different image types.
* Adds CTS tests for different image formats
* Defines the primary and optional supported image formats.
* Deletes redundant ERROR_IMAGE_FORMAT_NOT_SUPPORTED error code.
* Deletes UR_MEM_TYPE_BUFFER as it should be redundant to urMemBufferCreate entry-point.
* Deletes URMEM_TYPE_IMAGE1D_BUFFER which is only used in opencl and opencl will need cl_image_desc.buffer/mem_object value when CL_MEM_OBJECT_IMAGE1D_BUFFER is used, and we don't have a way in urMemImageCreate to pass it a buffer object that it will use to initialize an image with it and use this type. And if we initialized a new buffer inside the entry-point, we don't have a way of clearing the buffer memory at the end.
* Make sure the adapters return the right error-codes for urMemImageCreate entry-point.

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -506,10 +549,38 @@ urCommandBufferCreateExp(ur_context_handle_t Context, ur_device_handle_t Device,
                                    &ZeCommandListDesc, &ZeCommandList));
+
+  // Create a list for copy commands.
+  // Note that to synply the implementation, the current implementation only
```
```suggestion
  // Note that to simplify the implementation, the current implementation only
```

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -506,10 +549,38 @@ urCommandBufferCreateExp(ur_context_handle_t Context, ur_device_handle_t Device,
+  // Note that to synply the implementation, the current implementation only
+  // uses the main copy engine and does not use the link engine even if
+  // available.
+  ze_command_list_handle_t ZeCopyCommandList;
```
Not sure if we want to leave this uninitialized when there's not a copy engine
```suggestion
  ze_command_list_handle_t ZeCopyCommandList = nullptr;
```

EwanC commented on source/adapters/level_zero/command_buffer.hpp:
```diff
@@ -54,6 +56,15 @@ struct ur_exp_command_buffer_handle_t_ : public _ur_object {
+  // Level Zero Copy command list descriptor
+  ZeStruct<ze_command_list_desc_t> ZeCopyCommandListDesc;
+  // Indicates if a copy engine is available for use
+  bool MUseCopyEngine = false;
```
Not sure if we need a separate member variable for this, or we can just compare `ZeCopyCommandList` against `nullptr`

Bensuo replied:
Hmm, I sort of agree but I didn't really like the idea of having to replace the use of this variable with a nullptr check. Instead I've converted this bool to a bool returning function that checks the nullptr here. This makes the call sites a bit cleaner but also leaves it open to more complicated checks in the future if required which then don't have to be replicated at all call sites or refactored.

EwanC commented on source/adapters/level_zero/command_buffer.hpp:
```diff
@@ -54,6 +56,15 @@ struct ur_exp_command_buffer_handle_t_ : public _ur_object {
+  ZeStruct<ze_command_list_desc_t> ZeCopyCommandListDesc;
+  // Indicates if a copy engine is available for use
+  bool MUseCopyEngine = false;
+  // This flag is must be set to true if at least one copy command has been
```
```suggestion
  // This flag is must be set to false if at least one copy command has been
```

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -422,13 +446,35 @@ static ur_result_t enqueueCommandBufferFillHelper(
+                .ZeProperties.maxMemoryFillPatternSize < PatternSize) {
+      // If the maximum fill pattern size of the copy engine is smaller than the
+      // current PatternSize, the command is added to the compute command list.
+      ZeCommandList = CommandBuffer->ZeCommandList;
```
Should there still be an assert on the `maxMemoryFillPatternSize` of the compute queue here, like on line 470

Bensuo replied:
Yes there should, good catch.

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -1081,6 +1168,22 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+  // The Copy command-list is submitted to the main copy queue iif it is not
+  // empty.
+  if ((CommandBuffer->MUseCopyEngine) &&
+      (!CommandBuffer->MCopyCommandListEmpty)) {
```
Can we just check `!CommandBuffer->MCopyCommandListEmpty` in this if? For this to be set a copy engine command-list I think would have to exist

Bensuo replied:
Yes I think that's fine, I've removed that part of the condition.

EwanC reviewed (commented):
Rebasing this on tip UR would help review

EwanC commented:
@fabiomestre Pointed out that there's environment variables in https://github.com/oneapi-src/unified-runtime/issues/1528 that mention copy engines, and whether we should guard when we do this optimization based on those.

Bensuo commented:
> @fabiomestre Pointed out that there's environment variables in [oneapi-src#1528](https://github.com/oneapi-src/unified-runtime/issues/1528) that mention copy engines, and whether we should guard when we do this optimization based on those.

I don't think currently we respect any of these. From a quick look:

- `UR_L0_USE_COPY_ENGINE` - Needs some more investigation to understand what this does and how we would best support it, curently I think we'll only use the copy engine if there is a single main engine available.
- `UR_L0_USE_COPY_ENGINE_FOR_FILL` & `UR_L0_USE_COPY_ENGINE_FOR_D2D_COPY` - These we could respect pretty easily.
- `UR_L0_USE_COPY_ENGINE_FOR_IN_ORDER_QUEUE` - This one could be a bit confusing to respect, since the in-order-ness of the UR queue doesn't necessarily affect the command-buffer being in-order. It is enabled by default,  so perhaps we should make it the default also, but if we wanted to have this option I'd probably argue for a command buffer specific option. This is in contrast to the current behaviour in this PR which only uses copy engines for out-of-order command lists in command-buffers.

These do seem like mostly performance/debugging tweaks so I don't know that respecting them is super high priority, The Fill/D2D copy ones should be pretty trivial to implement though.

EwanC reviewed (approved):
LGTM

fabiomestre commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -544,6 +635,15 @@ urCommandBufferCreateExp(ur_context_handle_t Context, ur_device_handle_t Device,
+    // The copy command-list must be executed once the preconditions have been
+    // met. We therefore begin this command-list with a barrier on the
+    // preconditions.
+    ZE2UR_CALL(
```
Isn't this the same barrier as the one in line 635? Should the barrier be for ZeCopyCommandList?

Bensuo replied:
Good catch, fixed!

fabiomestre commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -1087,6 +1227,21 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+    auto &QGroupCopy = Queue->getQueueGroup(true);
+    uint32_t QueueGroupOrdinal;
+    auto &ZeCopyCommandQueue = QGroupCopy.getZeQueue(&QueueGroupOrdinal);
+    ZE2UR_CALL(zeFenceCreate, (ZeCopyCommandQueue, &ZeFenceDesc, &ZeCopyFence));
```
Why do we need to return a Fence here? I don't think we are synchronizing that fence anywhere right? 

I have similar comment about line 1227 but that's already existing code. I removed that fence in this PR with no immediately visible drawbacks: https://github.com/oneapi-src/unified-runtime/pull/1709/files

EwanC replied:
We're going to start using the fences in this PR https://github.com/oneapi-src/unified-runtime/pull/1629

fabiomestre replied:
I had forgotten about that PR. I think that makes the Fence for the compute command-list useful. 

Still unsure about the fence for the copy command-list. Sounds like it's dead code until we support updating memory operation nodes?

EwanC replied:
Good point, until we can update commands in a copy command-list I don't think we need to use the fence for this command-list in the same way inside our update implementation. So think we could omit creating and using an event here.

Bensuo replied:
I've removed the fence from the copy command list for now!

Bensuo committed:
[EXP][CMD-BUF][L0] Use copy engine to optimize cmd-buffer usage

- Out-of-order command-buffers now use copy-engine in some cases where available
- Also respect existing L0 env vars for using copy-engine with fill and d2d copies

kbenzie committed:
Merge pull request #1409 from omarahmed1111/Add-CTS-tests-for-image-format

[CTS] Add CTS tests for urMemImageCreate entry-point

Bensuo commented:
Closed in favor of upstream PR: https://github.com/oneapi-src/unified-runtime/pull/1738
